### ISSUE TITLE
Deleting CWD causes xonsh crash

### ISSUE THREAD

**Issue by astronouth7303** [xonsh/xonsh#1744]
If another process deletes the current working directory and then you run a command:

```
🐚 ls
Traceback (most recent call last):
  File "/usr/bin/xonsh", line 3, in <module>
    main()
  File "/usr/lib/python3.5/site-packages/xonsh/main.py", line 221, in main
    shell.shell.cmdloop()
  File "/usr/lib/python3.5/site-packages/xonsh/ptk/shell.py", line 119, in cmdloop
    self.default(line)
  File "/usr/lib/python3.5/site-packages/xonsh/base_shell.py", line 183, in default
    builtins.__xonsh_env__['PWD'] = os.getcwd()        # track any change in working directory
FileNotFoundError: [Errno 2] No such file or directory
```

And then xonsh crashes.

**Comment by laerus** [comment]
hey @astronouth7303, any idea which proc and how it deletes the cwd?

**Comment by laerus** [comment]
well just doing a `del $PWD` makes xonsh crash
```
Traceback (most recent call last):
  File "/home/laerus/.local/bin/xonsh", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/laerus/projects/xonsh/scripts/xonsh", line 3, in <module>
    main()
  File "/home/laerus/projects/xonsh/xonsh/main.py", line 221, in main
    shell.shell.cmdloop()
  File "/home/laerus/projects/xonsh/xonsh/ptk/shell.py", line 119, in cmdloop
    self.default(line)
  File "/home/laerus/projects/xonsh/xonsh/base_shell.py", line 188, in default
    self._fix_cwd()
  File "/home/laerus/projects/xonsh/xonsh/base_shell.py", line 224, in _fix_cwd
    if cwd != builtins.__xonsh_env__['PWD']:
  File "/home/laerus/projects/xonsh/xonsh/environ.py", line 824, in __getitem__
    raise KeyError(e.format(key))
KeyError: 'Unknown environment variable: $PWD'
```

**Comment by laerus** [comment]
i think we should not let the user delete envvars that are crucial to xonsh runtime

**Comment by laerus** [comment]
potential fix in  #1745

**Comment by astronouth7303** [comment]
In my case, it was another terminal. xonsh1 was in my project directory, and xonsh2 was deep in a build directory. xonsh1 cleans up the build directory out from under xonsh2, leaving xonsh2 floating in filesystem lala land.

I don't think your PR will fix it, because `os.getcwd()` immediately above it is throwing an error, probably because the PWD doesn't exist anymore. I believe the correct solution is to just skip the logic and print a warning if `FileNotFoundError` is thrown.

**Comment by laerus** [comment]
ooh okay i see what you mean, i thought for some reason you were talking about the `$PWD` envvar, i will get into that then and solve two issues at one go.

**Comment by laerus** [comment]
i'm not sure how we should go about fixing this, even if we catch the error on `base_shell._fix_cwd` and just return there are other places where `os.getcwd()` is used for example in `pyghooks`. Maybe we should drop the user in the home directory or smthn.

**Comment by astronouth7303** [comment]
Oh, yeah, ouch.

My preferred algorithm would do a string-wise crawl up the tree until an existing directory is found. (And on windows, whence you get to the root, try drive letters starting a `C:`.)

**Comment by laerus** [comment]
not sure if this would be a good idea, i think we should just print and error message and do nothing cause you may never want the command to execute on other/parent dirs.

**Comment by astronouth7303** [comment]
It's a better idea than dropping to `$HOME`, which has the potential to end up in the same situation.

Assuming non-catastrophic instances, climbing to the parent directory will likely land them logically in whatever they were working (eg, the project containing the generated directories, their home dir, etc.) In other cases, I don't think it's an unreasonable place

For catastrophic cases (`rm -rf` runs amok, unmounting on Windows (if it allows that)): Linux always has at least a root directory (I think); Windows will always have at least the system drive (because it actually locks files. But I think modern Windows may allow you to assign drive letters arbitrarily, so `C:` could potentially not exist (even though this configuration is extremely unlikely in the wild).

**Comment by laerus** [comment]
Yes dropping to home as i suggested is not a good idea either for the same reason, i believe we should just interrupt the command and print an error, cause we must mostly account for the most catastrophic case imo. I wouldn't like at all the case where `rm -rf` or w/e other crazy command i want to run doesn't actually land on the pwd and somewhere on the parent dirs (including my home dir).

**Comment by astronouth7303** [comment]
The problem with just printing and moving on is that there is very certainly other code that depends on `os.getcwd()`. (I didn't even know it verified the directory exists.) I'm very certain something else will fail and cause problems.

**Comment by laerus** [comment]
yeah it seems like that `os.getcwd()` runs each time we hit return, that's why i think we should never go beyond just checking pwd before doing anything.

**Comment by astronouth7303** [comment]
I mean, things like prompt generation and avox will use it and fail.

**Comment by scopatz** [comment]
Lesser shells don't even print an error message.  They won't let you cd into a dir that doesn't exist, but if you are in a dir that is removed underneath you, they'll chug along.  I think we need some behaviour like this.  I thought it worked like that already.  It sort of does, but it looks like when you run a command or try to cd out it crashes.  

It is suprising easy to remove dirs you are in with version control :)

**Comment by astronouth7303** [comment]
Normally, doing the expected thing for shells is goo, but it's `os.getwd()` that's throwing the error. Unless we mock it out, everywhere that calls it would have to handle the error in some way.

So instead of dealing with it in innumerable places, what's just deal with it in one.

**Comment by bobhy** [comment]
Changing working directory to some other location under the covers could cause a Xonsh script to, e.g, delete the wrong set of files.  That's a non-starter.

I was thinking that we could have Xonsh lock the CWD directory.  CMD.EXE does this on Windows
```
> rmdir /s xonsh\foo
xonsh\foo, Are you sure (Y/N)? y
The process cannot access the file because it is being used by another process.
```

I don't like the idea of adding code in Xonsh to cause an error to some other process.

How about letting Xonsh simply fail, but arrange to print an informative error at the next prompt?
That would be only one place to handle failure of `os.getcwd()`

**Comment by astronouth7303** [comment]
@bobhy No, because we're talking about doing it during pre-prompt clean up.

Assuming that xonsh itself does not crash--remember, xonsh crashing usually means the user never sees the backtrace that caused it to crash because the GUI terminal immediately closes--still leaves the user unable to safely do anything besides `cd`, and praying that no extensions break themselves.

**Comment by airdrik** [comment]
I've been running into this lately, as I've been cd-ing into project target/build directories to run the executables there.  I regularly do clean+build on the projects which will delete and recreate the target directory, oblivious to the fact that I might have a shell prompt in one of the target directories.
IMO, my order of priority for what needs fixed:
1: Do Not Crash!  Just because the current directory doesn't exist doesn't mean the world has ended.  Keep calm and carry on
2: Display a warning letting the user know that the current working directory has been updated
3: Stay in the same path if feasible.
3a: I wouldn't mind if it updated itself to the new target directory after I've clean+built, so that I don't have to do anything to continue working almost as if nothing happened.  In bash, you have to run "cd ." (change directory into the current directory) to accomplish this, so I'd be fine if that were all that is necessary.  But if the current directory being deleted and recreated could be auto-detected and handled (with a warning) that would be appreciated.
4: if 3 is not feasible, then I like the idea of walking up the directory tree to find a suitable existing directory.  An idea for the behavior here would be that if the user is trying to run a command when this triggers then update the directory, display a warning, don't execute the command and show the command that they tried to run so they can fix it as necessary and rerun.  
I think that given the need to change directory to another existing directory, the only circumstance that would allow for running the command would be to run the command with the (now deleted) previous working directory. Even then, chances are it will still fail or behave badly.  Presenting the command to the user to decide what to do with it seems reasonable enough to me.

**Comment by scopatz** [comment]
I agree that this needs to be fixed post-haste.  Even with @laerus's fixes in #1745, I find that this still happens.

**Comment by laerus** [comment]
yep the PR didnt fix this issue, the problem is that we call `os.getcwd()` and not catching the exception it raises when the dir was deleted, and there are potentially a lot of places where we ask about the cwd, for example when constructing the prompt.

**Comment by zzamboni** [comment]
I also encountered this problem today (xonsh 0.5.6). In my case I was in a network-mounted directory that got unmounted by another process, but I was able to reproduce the error by manually deleting the current directory from a different shell:
```
taazadi1@UM00104 ~/Downloads/foobar $ ls
    (in the meantime, delete "foobar" from another shell)
taazadi1@UM00104 ~/Downloads/foobar $ ls
Traceback (most recent call last):
  File "/usr/local/Cellar/xonsh/0.5.6/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 19735, in main
    return main_xonsh(args)
  File "/usr/local/Cellar/xonsh/0.5.6/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 19764, in main_xonsh
    shell.shell.cmdloop()
  File "/usr/local/Cellar/xonsh/0.5.6/libexec/lib/python3.6/site-packages/xonsh/ptk/shell.py", line 156, in cmdloop
    self.default(line)
  File "/usr/local/Cellar/xonsh/0.5.6/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 14843, in default
    self._fix_cwd()
  File "/usr/local/Cellar/xonsh/0.5.6/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 14877, in _fix_cwd
    cwd = os.getcwd()
FileNotFoundError: [Errno 2] No such file or directory
Xonsh encountered an issue during launch
Failback to /bin/bash
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
```

For reference, other shells produce an error, but don't crash when this happens (IMO, the shell should *NEVER* crash!):

fish:
```
taazadi1@UM00104➜ ~/D/foobar 
<E> fish: getcwd() failed with errno 2/No such file or directory
<E> fish: getcwd() failed with errno 2/No such file or directory
```

bash doesn't even notice until you try some file/dir operation:
```
bash-3.2$ ls -l
bash-3.2$ touch blah
touch: blah: No such file or directory
bash-3.2$ cd .
cd: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
```

**Comment by scopatz** [comment]
Hi @zzamboni - I think we all agree that xonsh should never crash. It is still on our to-do list, but of course, pull requests are always welcome!

**Comment by scopatz** [comment]
See #2266 for a potential fix.

**Comment by gforsyth** [comment]
Closed via #2266
